sw/host/opentitanlib/src/image/image.rs

```diff
@@ -102,120 +104,110 @@ impl Image {
-            .extensions
-            .entries
-            .iter()
-            .find(|x| x.identifier == id)
-            .ok_or(ImageError::NoExtensionTableEntry(id))?;
-        let offset = entry.offset as usize + std::mem::size_of::<ManifestExtHeader>();
-        let extension_slice = self
-            .data
-            .bytes
-            .get_mut(offset..offset + bytes.len())
-            .ok_or(ImageError::ExtensionOverflow)?;
-        extension_slice.copy_from_slice(bytes);
+    /// Adds an extension to the signed region of this `Image`.
+    pub fn add_signed_manifest_extensions(&mut self, spec: &ManifestExtSpec) -> Result<()> {
+        for entry_spec in &spec.signed_region {
+            self.add_manifest_extension(ManifestExtEntry::from_spec(
+                entry_spec,
+                spec.source_path(),
+            )?)?;
+        }
         Ok(())
     }
 
-    /// Adds a new extension at the end of the image and updates size.
-    /// Note that this method does NOT update the `signed_region_end` or `length` fields of the
-    /// manifest.
-    /// New extensions are ALWAYS added at the end of the image by design.
-    pub fn add_extension_data(&mut self, bytes: &[u8]) -> Result<()> {
-        let end_index = self
-            .size
-            .checked_add(bytes.len())
+    /// Adds an extension to the unsigned region of this `Image`.
+    pub fn add_unsigned_manifest_extensions(&mut self, spec: &ManifestExtSpec) -> Result<()> {
+        for entry_spec in &spec.unsigned_region {
+            self.add_manifest_extension(ManifestExtEntry::from_spec(
+                entry_spec,
+                spec.source_path(),
+            )?)?;
+        }
+        Ok(())
+    }
+
+    /// Adds an extension to the end of this `Image`.
+    pub fn add_manifest_extension(&mut self, entry: ManifestExtEntry) -> Result<()> {
+        let manifest = self.borrow_manifest()?;
+
+        // A copy of the extension table since we can't borrow as mutable.
+        let mut ext_table = manifest.extensions.entries;
+
+        let entry_id = entry.header().identifier;
+
+        // Update the offset in the extension table.
+        let ext_table_entry = ext_table
+            .iter_mut()
+            .find(|e| e.identifier == entry_id)
+            .ok_or(ImageError::NoExtensionTableEntry(entry_id))?;
+
+        // If the extension already exists, overwrite it, else append it to the end of the
+        // image.
+        let offset = if ext_table_entry.offset != 0 {
+            ext_table_entry.offset as usize
+        } else {
+            ensure!(
+                self.size % align_of::<u32>() == 0,
+                ImageError::BadExtensionAlignment(entry_id)
+            );
+            self.size
+        };
+        ext_table_entry.offset = offset as u32;
+
+        // Write the extension to the end of the image.
+        let ext_bytes = entry.to_vec();
+        let end_index = offset
+            .checked_add(ext_bytes.len())
```

Comment:
Good job using `checked_add` but can you verify that it's doing what we want? It looks like `offset` is `usize` not `u32`.